### PR TITLE
Try: Re-enable React StrictMode

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?

PR re-enable React StrictMode for editors. It was removed in #47701 and never restored.

## Why?
* Could be helpful with the React 19 upgrade.
* The StrictMode is mandatory for the React Compiler.

## Bugs Discovered
* https://github.com/WordPress/gutenberg/pull/62051
* https://github.com/WordPress/gutenberg/pull/62053
* https://github.com/WordPress/gutenberg/pull/62059
* https://github.com/WordPress/gutenberg/pull/62140
* https://github.com/WordPress/gutenberg/issues/62089
* https://github.com/WordPress/gutenberg/issues/62105

## Testing Instructions
* Smoke test the editors.
* CI checks should be green.

### Testing Instructions for Keyboard
Same.

### CODE REVIEW

**Comment by Mamaduka** [comment]
The e2e run with `SCRIPT_DEBUG` [2a60f04](https://github.com/WordPress/gutenberg/pull/61943/commits/2a60f04d43501ee13f05972c8c7d81e6431b872f) should give us a better idea about `StrictMode` and related issues.

File: .wp-env.json
Comment (youknowriad) [inline .wp-env.json]: I'm not entirely sure about this. I think it's good for the dev environment to have this enabled by I believe that ideally the tests environment should be as close to prod as possible.

Also, for performance tests, this will have an impact.

Comment (Mamaduka) [reply .wp-env.json]: This is temporary to see if anything fails in the development environment when StictMode is enabled.

**Review by youknowriad — APPROVED** [review]
Testing locally, I didn't see anything obvious. LGTM

**Comment by tyxla** [comment]
Thanks @Mamaduka!

I'd love to do some more thorough testing early next week, but if y'all are confident, go for it.

**Comment by Mamaduka** [comment]
As you can see, some e2e tests are failing when they're run in development and strict mode. This probably means we've subtle bugs in effects or anything that runs twice in `StrictMode`.

**Comment by tyxla** [comment]
> As you can see, some e2e tests are failing when they're run in development and strict mode. This probably means we've subtle bugs in effects or anything that runs twice in `StrictMode`.

Sounds familiar 😅

**Comment by Mamaduka** [comment]
Technically, I guess we can merge this since it doesn’t affect production apps.

Then create a separate tracking PR, where we try to fix each failure. Just like we do for React Compiler.

What do you think?

**Comment by youknowriad** [comment]
Yeah, since these bugs are actually already on trunk. I'm ok with merging.

The downside is that some folks might start noticing the bugs in their dev environments and wonder what's happening.

**Comment by youknowriad** [comment]
Actually, let's not merge this in 6.6 (unless we fix all the bugs).

**Comment by Mamaduka** [comment]
Sounds good. I’ll keep temp commit and we can use this as tracking PR.

**Comment by Mamaduka** [comment]
I started fixing the dev environment bugs. The "good news" is that some failures have related causes (e.g., #62053).

**Comment by ellatrix** [comment]
Is the list of bugs in the PR description complete?

**Comment by Mamaduka** [comment]
> Is the list of bugs in the PR description complete?

Not yet. I was trying to isolate and fix the ones that greatly affected the app, like the `useFocusOnMount` bug.

I'll try to complete it today.

**Comment by Mamaduka** [comment]
I think it's a complete list of bugs (caught by e2e tests). The tests report two false positives, which fail due to warnings generated in development mode but aren't real issues.

I'll create separate issues for the remaining problems.

**Comment by Mamaduka** [comment]
The only known bug that remains to be fixed is https://github.com/WordPress/gutenberg/pull/62140. I would appreciate review/testing.

**Comment by youknowriad** [comment]
@Mamaduka Are you certain about the link in your last comment?

**Comment by Mamaduka** [comment]
@youknowriad, it should be fixed now. GitHub autocorrected my PR number :/

**Comment by Mamaduka** [comment]
This is ready to be merged, but we will do it after this week's RC. It is probably not worth adding this as a minute change to WP  6.6.

**Review by tyxla — COMMENTED** [review]
Noting that at this time this needs to be rebased to take [this fix](https://github.com/WordPress/gutenberg/pull/62154) into consideration otherwise there are PHP fatals. 

I'm giving it a spin with a locally rebased branch to the latest `trunk`.

**Review by tyxla — COMMENTED** [review]
Do we also want to add strict mode to the reusable blocks?

https://github.com/WordPress/gutenberg/blob/00cfeee55e19be3f590748dd5d5f8a23c93d0a00/packages/list-reusable-blocks/src/index.js#L47

**Review by tyxla — APPROVED** [review]
Tested all editors, working pretty well. Shipping after the next release sounds sensible and will give us a good window to fix any incoming problems 👍

**Comment by Mamaduka** [comment]
> Noting that at this time this needs to be rebased to take https://github.com/WordPress/gutenberg/pull/62154 into consideration otherwise there are PHP fatals.

So, just rebase on top of the latest trunk, or are there any additional fixes required?

**Comment by tyxla** [comment]
> > Noting that at this time this needs to be rebased to take #62154 into consideration otherwise there are PHP fatals.
> 
> So, just rebase on top of the latest trunk, or are there any additional fixes required?

Yes, just rebasing fixes it.

**Review by kevin940726 — APPROVED** [review]
LGTM 👍. I'd expect to see warnings once this is merged but better sooner than later 😆.

**Comment by andrewserong** [comment]
It looks like this change caused a regression in the customizer's widget editor (reported in #62898). Just sharing here in case anyone has an idea for how to fix it? I wasn't sure if it was safe to revert the `<StrictMode>` in `customize-widgets/src/index.js` or if there's a better approach.

**Comment by t-hamano** [comment]
Strangely, two commands seem to no longer work when `StrictMode` is enabled. Please see #64004.

**Pull request merged**
